[rush-lib] Ignore specified files when calculating project state hash

elliot-nelson opened the pull request:
## Summary

Implementation of https://github.com/microsoft/rushstack/issues/2618.  The intent is to allow each project (in its `rush-project.json` file) to optionally specify glob patterns that should be ignored when calculating project state, which in turn, will allow those files to be ignored for the purposes of incremental build watching, cloud build cache hits, etc.

The new option is called `incrementalBuildIgnoredGlobs`, and is interpreted as a dot-ignore file.

## Details

PR needs a little cleanup, but it does work (tested locally on our monorepo here and does exactly what I was hoping).  Hoping to get feedback before I finish up the PR, to make sure this approach makes sense.

The new model of `ConfigurationFile`, which helps define riggable / extendable configuration files, is all `async`-based and takes a `Terminal` for possible warning output; this means that anything above it that uses it _also_ needs to be async and to accept a Terminal.  This requires a little bit of hacking on several chains (like `ProjectWatcher` -> `ProjectBuildCache` -> `PackageChangeAnalyzer` etc.).

It's worth noting that `ProjectBuildCache()` always accepted a `RushProjectConfiguration` that was previously loaded, but, that wasn't that useful in this situation.  That config is for the project for which a hash is being calculated; if, for example, `A` is dependent on `B`, if `B` specified `disableBuildCache` in its project configuration, that doesn't stop `A` from calculating the files in `B` as part of its hash.  But for this new feature, we _do_ need `A` to ignore files that `B` ignores, so we need to load the project configuration for every project in the repo essentially.  (Or, possibly, load them on the fly as we hit projects we haven't loaded the project configuration for yet; I opted for the former approach for simplicity but we can enhance that.)

## How it was tested

Currently tested in our current monorepo, for example, with project A depending on B:

 - Add `ignoreChangedFiles: "*.md"` to B's `rush-project.json`.
 - Tweak `B/src/index.ts` and `rush build`; rebuilds B and A as expected.
 - Tweak `B/README.md` and `rush build`; B and A get cache hits as expected.
 - Add `ignoreChangedFiles: "*.png"` to A's `rush-project.json`.
 - Tweak `A/logo.png` and `rush build`; A gets cache hit as expected.
 - Tweak `B/logo.png` and `rush build`; rebuilds B and A as expected (A is ignoring PNGs, but B is not)

<!--------------------------------------------------------------------------
👉 STEP 7: Don't forget to run "rush change":

     https://rushjs.io/pages/best_practices/change_logs/
--------------------------------------------------------------------------->

<!-- Have a question?  Ask for help in the chat room: https://rushstack.zulipchat.com/ -->

octogonz commented on apps/rush-lib/src/api/RushProjectConfiguration.ts:
```diff
@@ -21,6 +21,15 @@ interface IRushProjectJson {
+   * A list of file globs under the project root that will be ignored when
+   * determining whether this project has changed.
+   */
+  ignoreChangedFiles?: string[];
```
The phrase "changed files" sounds like it is somehow related to the `rush change` feature, which this isn't.

If they are globs, we should call them "globs".  (Heft has some settings like ["includeGlobs" and "excludeGlobs"](https://rushstack.io/pages/heft_tasks/copy-files/) for example.)

You suggested `ignoreFilePatternsForBuildState`, but that sounds limited to `rush build`, which this isn't.  The incremental build analysis can be enabled for various Rush **commands**, which includes `rush build` but also [custom commands](https://rushjs.io/pages/maintainer/custom_commands/).  A user might define an incremental command like `rush generate-docs` and then enable [this setting](https://rushjs.io/pages/configs/command-line_json/):

```
    //   /**
    //    * If true then this command will be incremental like the built-in "build" command
    //    */
    //   "incremental": false,
```
(We could provide separate ignore globs for each command, but that seems like an advanced feature that will get entangled with other design problems like [phased commands](https://github.com/microsoft/rushstack/issues/2300).)  My point is that "incremental" or "incremental build" is perhaps more clear than "build" by itself.

So based on all this, maybe the setting could be something like:
- `incrementalIgnoredGlobs` 
- `incrementalBuildIgnoredGlobs`
- `incrementalCommandIgnoredGlobs`

@iclanton what do you think?

elliot-nelson replied:
I think `incrementalBuildIgnoredGlobs` is just about right.  The phrase "incremental build" is the most common phrasing by far, in my opinion, so it makes sense to say "here's ignored globs for incremental build" (the fact that it applies to other commands potentially is OK, since those commands "act like incremental build" -- that's the way you'd describe them to someone).

octogonz commented on apps/rush-lib/src/schemas/rush-project.schema.json:
```diff
@@ -52,6 +52,15 @@
+
+    "ignoreChangedFiles": {
+      "type": "array",
+      "description": "A list of file globs under the project root that will be ignored when determining whether this project has changed.",
```
Possible improved wording:

> The incremental analyzer can skip Rush commands for projects whose input files have not changed since the last build.  Normally, every Git-tracked file under the project folder is assumed to be an input. Use the `incrementalBuildIgnoredGlobs` setting to ignore specific file paths. The paths are specified as globs relative to the project folder.

octogonz commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -71,17 +76,30 @@ export class PackageChangeAnalyzer {
+
+      const projectConfiguration:
+        | RushProjectConfiguration
+        | undefined = await RushProjectConfiguration.tryLoadForProjectAsync(project, undefined, terminal);
```
@iclanton Can the `rush-project.json` settings be inherited from a rig?  That seems useful

elliot-nelson replied:
I can confirm that they can today, I have it working in our local monorepo with 5.43.0.  🎉

dmichon-msft commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -71,17 +76,30 @@ export class PackageChangeAnalyzer {
+
+      const projectConfiguration:
+        | RushProjectConfiguration
+        | undefined = await RushProjectConfiguration.tryLoadForProjectAsync(project, undefined, terminal);
```
We'd want to do an async parallel invocation if we are waiting for file reads in this loop

dmichon-msft replied:
This function is very much critical path for all rush commands that do any work

elliot-nelson replied:
Makes sense! I was hunting around the codebase and I just might not have found it yet -- is there a generic bottleneck / rate limiting / parallelism-limited async map somewhere in rush-lib that I can use?

(I'm pretty sure dropping a `Promise.all(map)` bomb is actually slower for 200+ files, so I guess I'd feel better if we could limit it to a parallelism of 16 or something.)

octogonz replied:
>  I was hunting around the codebase and I just might not have found it yet -- is there a generic bottleneck / rate limiting / parallelism-limited async map somewhere in rush-lib that I can use?

@elliot-nelson Weirdly no, I can't seem to find one.  (Which is weird -- I'm sure someone encountered this before!)  The idiomatic pattern would be an API something like:

```js
import { Async } from '@rushstack/node-core-library';

// (or whatever the right API design is)
Async.throttle(3,
  [
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
  ] 
);
```

If you want to recommend a high quality throttling NPM package, we could use this PR to add a `@beta` API  in `@rushstack/node-core-library` that wraps it.

elliot-nelson replied:
I put up a PR with an alternate pretty quick option -- https://github.com/microsoft/rushstack/pull/2665.

dmichon-msft commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -92,7 +110,16 @@ export class PackageChangeAnalyzer {
         | undefined = this._rushConfiguration.findProjectForPosixRelativePath(filePath);
       if (owningProject) {
-        projectHashDeps.get(owningProject.packageName)!.set(filePath, fileHash);
+        const relativePath: string = path.relative(owningProject.projectRelativeFolder, filePath);
```
I'd recommend just slicing off the prefix, since `owningProject` will only be defined if filePath literally starts with `projectRelativeFolder`

dmichon-msft replied:
Also, it keeps all paths normalized to POSIX slashes (which git returns, and which projectRelativeFolder has been normalized to).

dmichon-msft commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -92,7 +110,16 @@ export class PackageChangeAnalyzer {
       if (owningProject) {
-        projectHashDeps.get(owningProject.packageName)!.set(filePath, fileHash);
+        const relativePath: string = path.relative(owningProject.projectRelativeFolder, filePath);
+        const ignoreMatchers: IMinimatch[] | undefined = ignorePatternsByProject.get(
```
We already have a dependency on `ignore`, which can run this like it is processing a .gitignore file and theoretically optimize a bit.

elliot-nelson commented:
Updated PR title and body, and made following changes:

 - Renamed option to `incrementalBuildIgnoredGlobs`, and updated the description based on @octogonz's suggestion.
 - Replaced `minimatch` with `ignore` (a great change). Added a unit test and slight wording change to convey this in the description -- this actually makes the feature _much more_ powerful, as it allows configurations like `['**', '!only-this-folder/*']`.  I don't know that we would necessarily highlight that fact, as a large monorepo with a bunch of fancy ignore logic is bound to start slowing down the incremental change logic.
 - Fixed up remaining unit tests and added rush change entry.

Took PR out of FEEDBACK state, as I think it is ready to go _except_ for the suggested improvement of loading `rush-project.json` files in parallel.  I can make that change if there's a good way in this codebase to load a limit of 10-16 or so files at a time.

elliot-nelson commented:
Latest cleanup commit -- load project configuration files in parallel, using the new `Async.forEachAsync` function that was ported over to the node core library.

At this point I believe all feedback has been addressed, let me know if any other concerns remain.

octogonz commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -29,9 +31,12 @@ export class PackageChangeAnalyzer {
   }
 
-  public getPackageDeps(projectName: string): Map<string, string> | undefined {
+  public async getPackageDeps(
```
This code uses a suffix to indicate async methods, so the name should probably be changed to `getPackageDepsAsync()`.  The same applies to other functions that were converted to async in this PR.

octogonz commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -157,6 +173,22 @@ export class PackageChangeAnalyzer {
+  ): Promise<Ignore> {
+    const projectConfiguration:
+      | RushProjectConfiguration
+      | undefined = await RushProjectConfiguration.tryLoadForProjectAsync(project, undefined, terminal);
```
(Ideally this should be cached somewhere, but we don't need to solve that in this PR.)

elliot-nelson replied:
I'm not sure we want to do that here; this is a wrapper for the `heft-config-file` library which does is own internal caching of all config files you load with it, and it seems like that's the best way to do it.

(I did file a related ticket, https://github.com/microsoft/rushstack/issues/2691, that will improve the caching further.)

octogonz commented on apps/rush-lib/src/logic/PackageChangeAnalyzer.ts:
```diff
@@ -92,7 +102,13 @@ export class PackageChangeAnalyzer {
-        projectHashDeps.get(owningProject.packageName)!.set(filePath, fileHash);
+        const relativePath: string = filePath
+          .replace(owningProject.projectRelativeFolder, '')
+          .replace(/^\//, '');
```
Could you add a code comment explaining what transformation is happening here?  

It might be helpful to include a realistic example of the input/output, something like:

```ts
        // Example:
        // filePath: category\project\src\thing.ts
        // projectRelativeFolder: category\project
        // relativePath: src\thing.ts
        const relativePath: string = filePath
          .replace(owningProject.projectRelativeFolder, '')
          .replace(/^\//, '');
```

It seems obvious to me now, but I had to run this code in the debugger to figure out what it is doing. :-)

octogonz replied:
The correctness of `replace()` hinges on the assumption that `filePath` will always contain `projectRelativeFolder`, but it's difficult to prove that, since `filePath` and `projectRelativeFolder` come from fairly different subsystems.

Here's a couple ideas to protect against that:
1. Use `path.relative()` to operate on path objects rather than text strings; OR
2. First test whether `filePath` starts with `projectRelativeFolder` and throw an `InternalError` if that not. The assumption is probably safe, but if not it would be nice to hear about it so we can fix it.

(BTW I tried some experiments to see if I could break this logic, like altering the case of a project folder on Windows. But the `LookupByPath` API malfunctions first preventing this code from running. Maybe I'm being overly defensive -- it's just if `replace()` did fail to match, the behavior might be pretty difficult to understand. 🙂)

@dmichon-msft

dmichon-msft replied:
The base logic here is that, since `LookupByPath` cannot return a matching project for a path that does not literally begin with `projectRelativeFolder`, all paths must necessarily begin with said string prefix. Personally I'd just blindly slice the expected length off of the start of the string, because that is how I prefer to do path manipulation (as quickly as possible, with a minimum of fuss).
Actually the `.slice()` call is safer here, because, contrary to what my memory had said, I only normalize the `projectRelativeFolder` to `/` when handing it off to `LookupByPath`, so you can technically (incorrectly) use `\\` inside of your paths in rush.json and weird things will happen.

elliot-nelson replied:
I've split the difference by adding a comment _and_ converting it to use @dmichon-msft 's suggested logic.

octogonz replied:
👍 I am content that sufficient paranoia has been applied here heheh

dmichon-msft commented:
In the repo I work on, this change has increased the run time of `PackageChangeAnalyzer._getData()` by an order of magnitude (from ~200 ms to 2.4 seconds). The bulk of this time appears to be in reading the rush-project.json files.
